### PR TITLE
Add io.mrarm.mcpelauncher

### PR DESCRIPTION
**Pull request by ChristopherHX**
Download, extract and play Minecraft from Google Play Store (if purchased) on Linux
~~Game launcher and Xbox live support are seperate packages.~~
~~Missing Appstream.~~

### CODE REVIEW

**Comment by ChristopherHX** [comment]
bot, build io.mrarm.mcpelauncher-ui

**Comment by ChristopherHX** [comment]
bot, build io.mrarm.mcpelauncher-ui

File: io.mrarm.mcpelauncher-ui.json [hunk 1]
```diff
@@ -0,0 +1,109 @@
+{
+    "id": "io.mrarm.mcpelauncher-ui",
+    "runtime": "org.kde.Platform",
+    "runtime-version": "5.12",
```
Comment (LeandroStanger) [inline io.mrarm.mcpelauncher-ui.json:4]: Update runtime to 5.14
```suggestion
    "runtime-version": "5.14",
```

Comment (ChristopherHX) [reply io.mrarm.mcpelauncher-ui.json]: Ok runtime updated

**Comment by ChristopherHX** [comment]
bot, build io.mrarm.mcpelauncher-ui

File: io.mrarm.mcpelauncher-ui.json (hunk 1 above)
Comment (ChristopherHX) [reply io.mrarm.mcpelauncher-ui.json]: But now qtwebengine isn't found.

Comment (ChristopherHX) [reply io.mrarm.mcpelauncher-ui.json]: This commit breaks kde runtime 5.13+ https://github.com/KDE/flatpak-kde-runtime/commit/fb5bbe40a1fc0406e101b1c9bfbe6262e0aa920d

**Comment by ChristopherHX** [comment]
bot, build io.mrarm.mcpelauncher-ui

**Comment by barthalion** [comment]
Some general feedback. I understand these are 3 separate components but Flatpak isn't a regular package manager to distribute software that way. First of all, running `flatpak-spawn` to execute another flatpak is a no-go, there's no way to express they depend on each other, not to mention it's virtually throws away any sandboxing Flatpak does. Ideally there should be just `io.mrarm.mcpelauncher`, with `msa-daemon` and `client` included; this way scripts are free to call each other without breaking promises Flatpak makes.

File: io.mrarm.mcpelauncher-ui.json [hunk 2]
```diff
@@ -0,0 +1,128 @@
+{
+    "id": "io.mrarm.mcpelauncher-ui",
+    "runtime": "org.kde.Platform",
+    "runtime-version": "5.14",
+    "sdk": "org.kde.Sdk",
+    "base": "io.qt.qtwebengine.BaseApp",
+    "base-version": "5.14",
+    "command": "mcpelauncher-ui-qt",
+    "build-options" : {
```
Comment (barthalion) [inline io.mrarm.mcpelauncher-ui.json:9]: SDK has its own CFLAGS set, it's better to use it instead.

File: io.mrarm.mcpelauncher-ui.json [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        "/bin/zipcmp",
+        "/bin/zipmerge",
+        "/bin/ziptool",
+        "/lib/libprotobuf-lite*",
+        "/lib/libprotoc*"
+    ],
+    "finish-args": [
+        "--socket=x11",
+        "--share=network",
+        "--filesystem=host",
```
Comment (barthalion) [inline io.mrarm.mcpelauncher-ui.json:26]: Why does it need full `host` access?

File: io.mrarm.mcpelauncher-ui.json [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+                }
+            ]
+        },
+        {
+            "name": "libzip",
+            "sources": [
+                {
+                    "type": "git",
+                    "url": "https://github.com/nih-at/libzip.git",
+                    "branch": "master"
```
Comment (barthalion) [inline io.mrarm.mcpelauncher-ui.json:47]: It should use specific commit or tag.

File: io.mrarm.mcpelauncher-ui.json [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+        "--talk-name=org.freedesktop.Flatpak"
+     ],
+    "modules": [
+        {
+            "name": "protobuf",
+            "sources": [
+                {
+                    "type": "git",
+                    "url": "https://github.com/protocolbuffers/protobuf.git",
+                    "branch": "3.12.x"
```
Comment (barthalion) [inline io.mrarm.mcpelauncher-ui.json:37]: It should use specific commit or tag.

File: io.mrarm.mcpelauncher-ui.json [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+            ],
+            "buildsystem": "cmake-ninja"
+        },
+        {
+            "name": "libuv",
+            "sources": [
+                {
+                    "type": "git",
+                    "url": "https://github.com/libuv/libuv.git",
+                    "branch": "v1.x"
```
Comment (barthalion) [inline io.mrarm.mcpelauncher-ui.json:58]: It should use specific commit or tag.

File: io.mrarm.mcpelauncher-ui.json [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+                }
+            ],
+            "buildsystem": "cmake-ninja"
+        },
+        {
+            "sources": [
+                {
+                    "type": "git",
+                    "url": "https://github.com/minecraft-linux/mcpelauncher-proprietary.git",
+                    "branch": "master"
```
Comment (barthalion) [inline io.mrarm.mcpelauncher-ui.json:79]: It should use specific commit or tag.

File: io.mrarm.mcpelauncher-ui.json [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+            ],
+            "buildsystem": "cmake-ninja"
+        },
+        {
+            "name": "mcpelauncher-ui",
+            "sources": [
+                {
+                    "type": "git",
+                    "url": "https://github.com/minecraft-linux/mcpelauncher-ui-manifest.git",
+                    "branch": "ng"
```
Comment (barthalion) [inline io.mrarm.mcpelauncher-ui.json:69]: It should use specific commit or tag.

File: io.mrarm.mcpelauncher-ui.metainfo.xml [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    <categories>
+        <category>Game</category>
+    </categories>
+
+    <recommends>
+        <id>io.mrarm.mcpelauncher-client</id>
+        <id>io.mrarm.msa-daemon</id>
+    </recommends>
+
+    <content_rating type="oars-1.1" />
```
Comment (barthalion) [inline io.mrarm.mcpelauncher-ui.metainfo.xml:35]: Passable for a launcher but it should describe Minecraft's rating.

File: io.mrarm.mcpelauncher-ui.json [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+                    "type": "file",
+                    "path": "io.mrarm.mcpelauncher-ui.metainfo.xml"
+                }
+            ],
+            "name": "compat",
+            "buildsystem": "simple",
+            "build-commands": [
+                "mkdir -p /app/bin/",
+                "echo '#!/bin/bash' > /app/bin/mcpelauncher-client",
+                "echo 'flatpak-spawn --host flatpak --filesystem=~/.var/app/$FLATPAK_ID:ro run app/io.mrarm.mcpelauncher-client// \"$@\"' >> /app/bin/mcpelauncher-client",
```
Comment (barthalion) [inline io.mrarm.mcpelauncher-ui.json:91]: As I said in separate comment, `flatpak-spawn --host flatpak` is a no-go.

File: io.mrarm.mcpelauncher-ui.json [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+                "chmod +x /app/bin/mcpelauncher-client",
+                "cd /app/share/applications && mv mcpelauncher-ui-qt.desktop io.mrarm.mcpelauncher-ui.desktop && sed -i 's/Icon=mcpelauncher-ui-qt/Icon=io.mrarm.mcpelauncher-ui/g' io.mrarm.mcpelauncher-ui.desktop",
```
Comment (barthalion) [inline io.mrarm.mcpelauncher-ui.json:93]: These should be installed under correct names by your build system instead of renamed and sed'ed here.

**Comment by barthalion** [comment]
I left some comments under the other two pull requests but comments posted here apply to them too.

File: io.mrarm.mcpelauncher-ui.json (hunk 11 above)
Comment (ChristopherHX) [reply io.mrarm.mcpelauncher-ui.json:93]: flatpak builder wants the applicationid be the filename, but the mcpelauncher-ui-qt submodule has it's own application id.
mcpelauncher-ui-qt isn't an valid flatpak name, but valid for all other builds.
Now added the right flatpak desktop file into the repo without using sed or renaming.

File: io.mrarm.mcpelauncher-ui.json (hunk 10 above)
Comment (ChristopherHX) [reply io.mrarm.mcpelauncher-ui.json:91]: Removed

File: io.mrarm.mcpelauncher-ui.metainfo.xml (hunk 9 above)
Comment (ChristopherHX) [reply io.mrarm.mcpelauncher-ui.metainfo.xml:35]: Added a more detailed rating.
Althought the game has ingame purchases on google play, you cannot buy the ingame coins with this launcher.

File: io.mrarm.mcpelauncher-ui.json (hunk 2 above)
Comment (ChristopherHX) [reply io.mrarm.mcpelauncher-ui.json:9]: Ok, the examples of flatpak had set them for whatever reasons.
Only the x86_64 gamelauncher don't compile with the default cflags of the sdk at all, clang-9, throws errors

File: io.mrarm.mcpelauncher-ui.json (hunk 3 above)
Comment (ChristopherHX) [reply io.mrarm.mcpelauncher-ui.json]: The file picker don't need it, removed now.

**Comment by barthalion** [comment]
bot, build io.mrarm.mcpelauncher-ui

**Comment by barthalion** [comment]
bot, build io.mrarm.mcpelauncher

File: flathub.json [hunk 12]
```diff
@@ -0,0 +1,4 @@
+{
+    "only-arches": [ "x86_64", "i386", "arm" ],
+    "skip-appstream-check": true
```
Comment (barthalion) [inline flathub.json:3]: Why did you disable appstream validation?

File: io.mrarm.mcpelauncher.json [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+    ],
+    "finish-args": [
+        "--socket=x11",
+        "--share=network",
+        "--device=dri",
+        "--socket=pulseaudio",
+        "--device=all"
+    ],
+    "modules": [
+        "mcpelauncher-manifest-deps.json",
```
Comment (barthalion) [inline io.mrarm.mcpelauncher.json:48]: Nitpick: storing all modules in the main manifest is usually preferred. The recommended key order is that `sources` is the last as it's not really significant for reading. It's your choice though.

File: io.mrarm.mcpelauncher.metainfo.xml [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+<component type="desktop-application">
+    <id>io.mrarm.mcpelauncher</id>
+    <metadata_license>CC0-1.0</metadata_license>
+    <project_license>GPL-3.0</project_license>
+    <name>Minecraft Bedrock Launcher</name>
+    <summary>Download and play Minecraft Bedrock Edition on GNU/Linux</summary>
+
+    <description>
+        <p>
+            The project runs the native libraries from the Android version directly on your computer. This is accomplished by fixing the incompatibilities between the libc used on Android and the one used on desktop Linux or OS X (Android - Bionic; Linux - glibc). This is a simple compatibility layer which doesn’t impact performance in any significant way; it’s similar to Wine, but much more lightweight and simple. Aditionally, all Android-specific code has been rewritten to run on Linux (AppPlatform, Store, Xbox Live, etc.).
```
Comment (barthalion) [inline io.mrarm.mcpelauncher.metainfo.xml:11]: Please add disclaimer like the one in the app that it's not unofficial and not affiliated with MS/Mojang. For example: https://github.com/flathub/com.slack.Slack/blob/master/com.slack.Slack.appdata.xml#L18

File: io.mrarm.mcpelauncher.metainfo.xml [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+
+    <launchable type="desktop-id">io.mrarm.mcpelauncher.desktop</launchable>
+
+    <screenshots>
+        <screenshot type="default">
+        <caption>Main screen of the Launcher UI</caption>
+        <image>https://user-images.githubusercontent.com/24195572/71060819-64f78c80-211b-11ea-8ea4-f1c2dabb9b54.png</image>
+        </screenshot>
+    </screenshots>
+    <categories>
```
Comment (barthalion) [inline io.mrarm.mcpelauncher.metainfo.xml:23]: It will be taken from desktop file, no need to define it here.

**Comment by barthalion** [comment]
Looks like `mcpelauncher-ui-qt` stays in the background even after I close the window. It'd nice to have it fixed too.

File: flathub.json (hunk 12 above)
Comment (ChristopherHX) [reply flathub.json]: By mistake, my two closed PR's have bad appstream files, no icon etc.

File: io.mrarm.mcpelauncher.metainfo.xml (hunk 14 above)
Comment (ChristopherHX) [reply io.mrarm.mcpelauncher.metainfo.xml:11]: I have added the discaminar from the launcher welcome page to the metadata

**Comment by ChristopherHX** [comment]
> `mcpelauncher-ui-qt` stays running in the background

0.0.2 should fix this bug for the welcome window

**Comment by ChristopherHX** [comment]
bot, build io.mrarm.mcpelauncher

Hope that this compiles..

File: io.mrarm.mcpelauncher.json (hunk 13 above)
Comment (ChristopherHX) [reply io.mrarm.mcpelauncher.json]: > storing all modules in the main manifest is usually preferred.
Ok now is all again in one json file.
source moved to bottom, but only-arches is still below the source.

**Comment by ChristopherHX** [comment]
bot, build io.mrarm.mcpelauncher

**Comment by ChristopherHX** [comment]
bot, build io.mrarm.mcpelauncher

**Comment by ChristopherHX** [comment]
bot, build io.mrarm.mcpelauncher

**Comment by ChristopherHX** [comment]
The arm flatpak is broken on a raspberry Pi2.
- libfmod fails to connect to pulseaudio via alsa (pulse_connect) on raspbian 10
- after merging manifests arm now fails to find libfmod

**Comment by ChristopherHX** [comment]
bot, build io.mrarm.mcpelauncher
Needs more testing

**Comment by ChristopherHX** [comment]
bot, build io.mrarm.mcpelauncher

**Comment by barthalion** [comment]
I don't own Minecraft so I cannot test it properly, but the launcher properly exits now so looks good at least to that point.

**Pull request closed without merging**